### PR TITLE
CartesianGrid reads chartWidth and chartHeight from context

### PR DESCRIPTION
**Pull request by PavelVanecek**
## Description

Another smallish change to the CartesianGrid

## Related Issue

https://github.com/recharts/recharts/discussions/3717

## Motivation and Context

No more element cloning

## How Has This Been Tested?

npm test

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes
- [X] All new and existing tests passed.

### CODE REVIEW

File: src/cartesian/CartesianGrid.tsx [hunk 1]
```diff
@@ -44,20 +45,26 @@ export type VerticalCoordinatesGenerator = (
...
 interface CartesianGridProps extends InternalCartesianGridProps {
+  /**
+   * The x-coordinate of grid.
+   * If left undefined, it will be computed from the chart's offset and margins.
+   */
+  x?: number;
+  /**
+   * The y-coordinate of grid.
+   * If left undefined, it will be computed from the chart's offset and margins.
+   */
```
Comment (nikolasrieble) [inline src/cartesian/CartesianGrid.tsx:66]: Nice, thank you for also adding the comments! ❤️

File: test/cartesian/CartesianGrid.spec.tsx [hunk 2]
```diff
@@ -30,6 +30,179 @@ describe('<CartesianGrid />', () => {
...
+            width={100}
+            height={200}
+            fill="green"
+            verticalPoints={verticalPoints}
+            horizontalPoints={horizontalPoints}
+          />
+        </Surface>,
+      );
+      const horizontalLines = container.querySelectorAll('.recharts-cartesian-grid-horizontal line');
+      expect(horizontalLines).toHaveLength(5);
```
Comment (nikolasrieble) [inline test/cartesian/CartesianGrid.spec.tsx:87]:
```suggestion
      expect(horizontalLines).toHaveLength(horizontalPoints.length);
```

File: test/cartesian/CartesianGrid.spec.tsx [hunk 3]
```diff
@@ -30,6 +30,179 @@ describe('<CartesianGrid />', () => {
...
+      horizontalLines.forEach(line => {
+        expect.soft(line).toHaveAttribute('x', '1');
+        expect.soft(line).toHaveAttribute('y', '2');
+        expect.soft(line).toHaveAttribute('x1', '1');
+        expect.soft(line).toHaveAttribute('x2', '101');
+      });
+      const verticalLines = container.querySelectorAll('.recharts-cartesian-grid-vertical line');
+      expect(verticalLines).toHaveLength(4);
```
Comment (nikolasrieble) [inline test/cartesian/CartesianGrid.spec.tsx:95]:
```suggestion
      expect(verticalLines).toHaveLength(verticalPoints.length);
```

File: test/cartesian/CartesianGrid.spec.tsx (hunk 2 above)
Comment (nikolasrieble) [reply test/cartesian/CartesianGrid.spec.tsx:87]: Refering the test data directly would make the test more robust, since it would also succeed if the data was adapted, right?

File: test/cartesian/CartesianGrid.spec.tsx [hunk 4]
```diff
@@ -30,6 +30,179 @@ describe('<CartesianGrid />', () => {
...
+  describe('layout and size when inherited from parent chart', () => {
+    const exampleMargin: Margin = {
+      bottom: 1,
+      left: 2,
+      right: 4,
+      top: 8,
+    };
+    it('should put x, y, width and height as coordinates to the background', () => {
+      const { container } = render(
+        <AreaChart margin={exampleMargin} width={500} height={400}>
```
Comment (nikolasrieble) [inline test/cartesian/CartesianGrid.spec.tsx:144]: I would suggest to always use the ComposedChart, since it is the most versatile. Or, we could make this a parametrised tests and simply test for all charts?

File: test/cartesian/CartesianGrid.spec.tsx [hunk 5]
```diff
@@ -30,6 +30,179 @@ describe('<CartesianGrid />', () => {
...
+    top: 14,
+  };
+
+  const expectedOffset: ChartOffset = {
+    bottom: 11,
+    brushBottom: 11,
+    height: 175,
+    left: 12,
+    right: 13,
+    top: 14,
```
Comment (nikolasrieble) [inline test/cartesian/CartesianGrid.spec.tsx:46]: What would happen if these numbers were not to align with the chartMargin? If the alignment is required, why not just reuse it directly? 
```suggestion
   ...chartMargin
```

File: test/cartesian/CartesianGrid.spec.tsx [hunk 6]
```diff
@@ -30,6 +30,179 @@ describe('<CartesianGrid />', () => {
...
+  describe('layout and size when set explicitly', () => {
+    it('should put x, y, width and height as coordinates to the background', () => {
+      const { container } = render(
+        <Surface width={500} height={500}>
+          <CartesianGrid
+            x={1}
+            y={2}
+            width={100}
+            height={200}
+            fill="green"
```
Comment (nikolasrieble) [inline test/cartesian/CartesianGrid.spec.tsx:59]:
```suggestion
```

Comment (nikolasrieble) [reply test/cartesian/CartesianGrid.spec.tsx:59]: This seems unrelated to the test. I believe it would help to keep tests as simple as possible to make them easy to understand.

File: test/cartesian/CartesianGrid.spec.tsx [hunk 7]
```diff
@@ -30,6 +30,179 @@ describe('<CartesianGrid />', () => {
...
+    });
+
+    it('should put x, y, width and height as coordinates to all lines', () => {
+      const { container } = render(
+        <Surface width={500} height={500}>
+          <CartesianGrid
+            x={1}
+            y={2}
+            width={100}
+            height={200}
```
Comment (nikolasrieble) [inline test/cartesian/CartesianGrid.spec.tsx:79]: It seems these values are set in every test. Should we also extract them into a shared object?, i.e. 
```suggestion
            ...DEFAULT_TEST_PARAMETERS
```

File: test/cartesian/CartesianGrid.spec.tsx [hunk 8]
```diff
@@ -30,6 +30,179 @@ describe('<CartesianGrid />', () => {
...
+            fill="green"
+            verticalPoints={verticalPoints}
+            horizontalPoints={horizontalPoints}
+            horizontalFill={['red', 'black']}
+            verticalFill={['green', 'blue']}
```
Comment (nikolasrieble) [inline test/cartesian/CartesianGrid.spec.tsx:116]: The fill seems unrelated to the test case.

File: test/cartesian/CartesianGrid.spec.tsx [hunk 9]
```diff
@@ -30,6 +30,179 @@ describe('<CartesianGrid />', () => {
...
+    });
+  });
+
+  describe('layout and size when inherited from parent chart', () => {
+    const exampleMargin: Margin = {
+      bottom: 1,
+      left: 2,
+      right: 4,
+      top: 8,
+    };
```
Comment (nikolasrieble) [inline test/cartesian/CartesianGrid.spec.tsx:141]: Why do we use a different exampleMargin here? Why not just the top level chartMargin?

File: test/cartesian/CartesianGrid.spec.tsx [hunk 10]
```diff
@@ -30,6 +30,179 @@ describe('<CartesianGrid />', () => {
...
+      const horizontalStripes = container.querySelectorAll('.recharts-cartesian-gridstripes-horizontal rect');
+      expect(horizontalStripes).toHaveLength(5);
+      horizontalStripes.forEach(stripe => {
+        expect.soft(stripe).toHaveAttribute('x', '2');
+        expect.soft(stripe).toHaveAttribute('width', '494');
+      });
+      const verticalStripes = container.querySelectorAll('.recharts-cartesian-gridstripes-vertical rect');
+      expect(verticalStripes).toHaveLength(5);
+      verticalStripes.forEach(stripe => {
+        expect.soft(stripe).toHaveAttribute('y', '8');
```
Comment (nikolasrieble) [inline test/cartesian/CartesianGrid.spec.tsx:200]: Interesting. Would you like to suggest us to use expect.soft always? 
Should we document this as a recommendation? 

I have seen this for the first time, but it does make sense. https://vitest.dev/api/expect.html#soft

File: test/cartesian/CartesianGrid.spec.tsx [hunk 11]
```diff
@@ -310,28 +483,26 @@ describe('<CartesianGrid />', () => {
           ticks: ['x', 'y', 'x'],
         };
         render(
-          <Surface width={500} height={500}>
+          <AreaChart width={300} height={200} margin={chartMargin}>
```
Comment (nikolasrieble) [inline test/cartesian/CartesianGrid.spec.tsx:486]: So, is this a breaking change? This case no longer works with the Surface?

File: test/cartesian/CartesianGrid.spec.tsx [hunk 12]
```diff
@@ -430,34 +587,27 @@ describe('<CartesianGrid />', () => {
...
-
           expect(horizontalCoordinatesGenerator).toHaveBeenCalledOnce();
           expect(horizontalCoordinatesGenerator).toHaveBeenCalledWith(
             {
-              yAxis: yAxisExpected,
+              yAxis: undefined,
               width: 300,
               height: 200,
-              offset,
+              offset: { bottom: 11, brushBottom: 11, height: 175, left: 12, right: 13, top: 14, width: 275 },
```
Comment (nikolasrieble) [inline test/cartesian/CartesianGrid.spec.tsx:610]: We have extract a constant in the top `expectedOffset`. Should we also use it here?

**Review by nikolasrieble — COMMENTED** [review]
Lots of questions, only on testing.

File: test/cartesian/CartesianGrid.spec.tsx (hunk 4 above)
Comment (PavelVanecek) [reply test/cartesian/CartesianGrid.spec.tsx:144]: This was easier than I expected

File: test/cartesian/CartesianGrid.spec.tsx (hunk 5 above)
Comment (PavelVanecek) [reply test/cartesian/CartesianGrid.spec.tsx:46]: I want this to be a little bit explicit, and I want the tests to fail when someone updates the margin but does not update the offset, and vice versa. If we DRY then people might miss the relationship between margin and offset here.

File: test/cartesian/CartesianGrid.spec.tsx (hunk 6 above)
Comment (PavelVanecek) [reply test/cartesian/CartesianGrid.spec.tsx:59]: This is required because without `fill` property, CartesianGrid does not render the background element

File: test/cartesian/CartesianGrid.spec.tsx (hunk 8 above)
Comment (PavelVanecek) [reply test/cartesian/CartesianGrid.spec.tsx:116]: This is required because without `horizontalFill` and `verticalFill`, CategoricalChart will not render the stripes

File: test/cartesian/CartesianGrid.spec.tsx (hunk 6 above)
Comment (PavelVanecek) [reply test/cartesian/CartesianGrid.spec.tsx:59]: But it's not required in the other two tests so I removed it from there

File: test/cartesian/CartesianGrid.spec.tsx (hunk 10 above)
Comment (PavelVanecek) [reply test/cartesian/CartesianGrid.spec.tsx:200]: I find it useful because sometimes I want to see all failures in a test to have a full picture of what is going on - it helps me debug.

I think it's a personal preference so I wouldn't make any recommendations. The tool is there, use it if you like.

File: test/cartesian/CartesianGrid.spec.tsx (hunk 11 above)
Comment (PavelVanecek) [reply test/cartesian/CartesianGrid.spec.tsx:486]: The CategoricalChart will no longer render on its own, it now requires a context coming from one of the generateCategoricalChart.

Do you think it's a breaking change? It should be fairly easy to fix (few null checks will do the trick), I just never expected that to be a use-case.

File: test/cartesian/CartesianGrid.spec.tsx (hunk 12 above)
Comment (PavelVanecek) [reply test/cartesian/CartesianGrid.spec.tsx:610]: Done

File: test/cartesian/CartesianGrid.spec.tsx (hunk 7 above)
Comment (PavelVanecek) [reply test/cartesian/CartesianGrid.spec.tsx:79]: Done! This was lot less easy :(

File: test/cartesian/CartesianGrid.spec.tsx (hunk 11 above)
Comment (PavelVanecek) [reply test/cartesian/CartesianGrid.spec.tsx]: Well actually it still does render independently (or with Surface, to get the svg tag parent) in this pull request. But it will break in the next pull request.

Comment (nikolasrieble) [reply test/cartesian/CartesianGrid.spec.tsx]: I agree that this use case is not intended, but I want to raise awareness to breaking changes. If this tests fails with the surface, it strictly speaking is a breaking change - that being said: One that I am willing to accept for moving forward with improved state management. The benefits do outweigh the cost imho. 

But others might disagree. @ckifer

Comment (ckifer) [reply test/cartesian/CartesianGrid.spec.tsx]: Hmm, not sure how I feel. The components that don't need context should probably render without it if all the right data is supplied. I could see people rendering things separately or doing some sort of really custom thing with the recharts building block components. 

But I agree in principle, we should move forward else we'll just keep existing where we are

**Review by ckifer — APPROVED** [review]
PR looks good overall, one question.

re: breaking change. Lets ship it and keep moving, I don't want to stunt momentum by overthinking things. If we get tons of reports after next release we can address

File: test/cartesian/CartesianGrid.spec.tsx (hunk 8 above)
Comment (nikolasrieble) [reply test/cartesian/CartesianGrid.spec.tsx:120]: Thank you for the context, I was not aware. This is helpful knowledge. ❤️

File: src/cartesian/CartesianGrid.tsx [hunk 13]
```diff
@@ -44,20 +45,26 @@ export type VerticalCoordinatesGenerator = (
...
-  chartWidth?: number;
-  chartHeight?: number;
 }
 
 interface CartesianGridProps extends InternalCartesianGridProps {
+  /**
+   * The x-coordinate of grid.
+   * If left undefined, it will be computed from the chart's offset and margins.
+   */
+  x?: number;
```
[the remark this answers is not in the document]
Comment (PavelVanecek) [reply src/cartesian/CartesianGrid.tsx:62]: They are documented on the Recharts.org website, that makes them external in my opinion.

File: test/cartesian/CartesianGrid.spec.tsx (hunk 11 above)
Comment (PavelVanecek) [reply test/cartesian/CartesianGrid.spec.tsx]: Does our website or examples show or promote usage of individual components? Most of them won't work alone anyway because of their dependency on generateCategoricalChart state. So I assumed that's a design decision.

If we do want to support individual rendering: we will document and export the ChartLayoutContext. Then we get maximum flexibility and anyone can rendering everything without gCC.

Perhaps we should start another discussion and see if that's valuable for our users?

File: src/cartesian/CartesianGrid.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (ckifer) [reply src/cartesian/CartesianGrid.tsx:62]: Got it. Some of them are denote "this is usually set internally", but agreed that external options are external options

[the remark this answers is not in the document]
Comment (PavelVanecek) [reply src/cartesian/CartesianGrid.tsx:62]: Most of everything is set internally :( I think at some point we will want to revisit our external props and get rid of some that don't make sense. Or, at least mark them as deprecated, with a warning or something

[the remark this answers is not in the document]
Comment (ckifer) [reply src/cartesian/CartesianGrid.tsx:62]: agreed

File: test/cartesian/CartesianGrid.spec.tsx (hunk 11 above)
Comment (ckifer) [reply test/cartesian/CartesianGrid.spec.tsx]: So I know that people have used shapes on their own (Rectangle, Curve, etc.). `Text` probably too.

Others that I'm not sure about but may be useful to render without a greater context:

* Legend
* Label (?) 

Everything else is probably ok as long as they do render _individually_ within a chart component.

ex: I want this to render an XAxis without a chart

```
<ComposedChart height={200} width={200}>
  <XAxis {...props I need for axis to render} >
</ComposedChart
```
